windows-driver-docs-pr/display/ccd-example-code.md

```diff
@@ -21,47 +21,215 @@ ms.technology: windows-devices
+			if (pathInfoArr[i].sourceInfo.id != primaryDisplay->id) {
+				pathInfoArr[i].flags |= DISPLAYCONFIG_PATH_ACTIVE;
+				pathInfoArr[i].sourceInfo.id = primaryDisplay->id;
+				pathInfoArr[i].sourceInfo.modeInfoIdx = DISPLAYCONFIG_PATH_MODE_IDX_INVALID;
+				pathInfoArr[i].targetInfo.modeInfoIdx = DISPLAYCONFIG_PATH_MODE_IDX_INVALID;
+			}
+		}
+		return didSucceed(SetDisplayConfig(numPathInfoArrayElements, pathInfoArr, numModeInfoArrayElements, modeInfoArr, 
+			SDC_APPLY | SDC_USE_SUPPLIED_DISPLAY_CONFIG | SDC_ALLOW_CHANGES | SDC_SAVE_TO_DATABASE));
+	}
+	return false;
+}
+
+// Returns the source device name for the given adapterId and id
+std::wstring getSourceDeviceName(LUID adapterId, UINT32 id) {
+	DISPLAYCONFIG_SOURCE_DEVICE_NAME deviceName;
+	DISPLAYCONFIG_DEVICE_INFO_HEADER header;
+	header.adapterId = adapterId;
+	header.id = id;
+	header.size = sizeof(DISPLAYCONFIG_SOURCE_DEVICE_NAME);
+	header.type = DISPLAYCONFIG_DEVICE_INFO_GET_SOURCE_NAME;
+	deviceName.header = header;
+	LONG status = DisplayConfigGetDeviceInfo((DISPLAYCONFIG_DEVICE_INFO_HEADER*)&deviceName);
+	if (didSucceed(status)) {
+		return std::wstring(deviceName.viewGdiDeviceName);
+	}
+	return L"";
+}
+
+std::wstring getSourceDeviceName(DISPLAYCONFIG_PATH_INFO pathInfo) {
+	return getSourceDeviceName(pathInfo.sourceInfo.adapterId, pathInfo.sourceInfo.id);
+}
+
+std::wstring getSourceDeviceName(DISPLAYCONFIG_MODE_INFO pathInfo) {
+	return getSourceDeviceName(pathInfo.adapterId, pathInfo.id);
+}
+
+// Returns the monitor friendly device name.
+// AdapterId and Id should both be from a target display, not a source.
+// (e.g. DISPLAYCONFIG_PATH_INFO.targetInfo or 
+//  DISPLAYCONFIG_MODE_INFO.infoType == DISPLAYCONFIG_MODE_INFO_TYPE_TARGET)
+std::wstring getTargetFriendlyDeviceName(LUID adapterId, UINT32 id) {
+	DISPLAYCONFIG_TARGET_DEVICE_NAME deviceName;
+	DISPLAYCONFIG_DEVICE_INFO_HEADER header;
+	header.adapterId = adapterId;
+	header.id = id;
+	header.size = sizeof(DISPLAYCONFIG_TARGET_DEVICE_NAME);
+	header.type = DISPLAYCONFIG_DEVICE_INFO_GET_TARGET_NAME;
+	deviceName.header = header;
+	LONG status = DisplayConfigGetDeviceInfo((DISPLAYCONFIG_DEVICE_INFO_HEADER*)&deviceName);
+	if (didSucceed(status)) {
+		return std::wstring(deviceName.monitorFriendlyDeviceName);
+	}
+	return L"";
+}
+
+std::wstring getTargetFriendlyDeviceName(DISPLAYCONFIG_PATH_INFO pathInfo) {
+	return getTargetFriendlyDeviceName(pathInfo.targetInfo.adapterId, pathInfo.targetInfo.id);
+}
+
+std::wstring getTargetFriendlyDeviceName(DISPLAYCONFIG_MODE_INFO pathInfo) {
+	return getTargetFriendlyDeviceName(pathInfo.adapterId, pathInfo.id);
+}
+
+int __cdecl main(int argc, const char* argv[]) {
+	// Get information on how many display configuration items there are
+	UINT32 numPathArrayElements = 0;
+	UINT32 numModeInfoArrayElements = 0;
+	LONG status = GetDisplayConfigBufferSizes(QDC_DATABASE_CURRENT, &numPathArrayElements, &numModeInfoArrayElements);
```

Comment:
Moreover, the mechanism used by this code to clone monitors needs to be updated for Windows 10 and virtual modes/topologies. The way it's written right now, it will find the source ID of the primary path and then find other paths and makes those have the same source ID. Note that this has a bug already because it's not comparing the adapter ID as well, so it has undefined behavior with multiple GPUs.

For Windows 10, virtual topologies can be used to clone paths based only on their x,y position. In this case, source IDs are actually different and only the position is used. It's a bit more complicated than this as well because each target can have virtual modes disabled/enabled individually, so if you are trying to clone a target that has virtual modes disabled, you can only clone it with other targets on the same adapter that also have virtual modes disabled. If it has virtual modes enabled, then you can clone it with all targets that have virtual modes enabled on all adapters.